File: ml-service/test_aquifer_integration.py
# Test script to verify aquifer integration
import requests
import json

def test_aquifer_integration():
    ML_SERVICE_URL = "http://localhost:8000"
    
    # Test data with Bangalore coordinates
    test_data = {
        "roof_area": 100.0,
        "roof_type": "concrete",
        "soil_type": "loamy",
        "annual_rainfall": 970.0,
        "state": "Karnataka",
        "district": "Bangalore Urban",
        "latitude": 12.9716,
        "longitude": 77.5946
    }
    
    print("🧪 Testing Aquifer Integration...")
    print(f"ML Service URL: {ML_SERVICE_URL}")
    print(f"Test Location: Bangalore ({test_data['latitude']}, {test_data['longitude']})")
    print("-" * 60)
    
    try:
        # Test 1: Direct aquifer info endpoint
        print("\n1️⃣  Testing Direct Aquifer Endpoint...")
        response = requests.get(f"{ML_SERVICE_URL}/aquifer-info/{test_data['latitude']}/{test_data['longitude']}")
        
        if response.status_code == 200:
            aquifer_data = response.json()
            print("✅ Aquifer endpoint works!")
            print(f"   - Success: {aquifer_data.get('success')}")
            if aquifer_data.get('success') and aquifer_data.get('nearest_aquifer'):
                print(f"   - Nearest Aquifer: {aquifer_data['nearest_aquifer']['name']}")
                print(f"   - Distance: {aquifer_data.get('distance_km')} km")
                print(f"   - Type: {aquifer_data.get('aquifer_type')}")
                print(f"   - Feasibility Score: {aquifer_data.get('feasibility_score')}/100")
        else:
            print(f"❌ Aquifer endpoint failed: {response.status_code}")
            print(f"   Response: {response.text}")
            
    except Exception as e:
        print(f"❌ Direct aquifer test failed: {str(e)}")
    
    try:
        # Test 2: Prediction endpoint with aquifer integration
        print("\n2️⃣  Testing Prediction with Aquifer Integration...")
        response = requests.post(f"{ML_SERVICE_URL}/predict", json=test_data)
        
        if response.status_code == 200:
            prediction_data = response.json()
            print("✅ Prediction endpoint works!")
            print(f"   - Potential Harvest: {prediction_data.get('potential_harvest')} L/year")
            print(f"   - Feasibility: {prediction_data.get('feasibility')}")
            print(f"   - Groundwater Level: {prediction_data.get('groundwater_level')}m")
            
            if prediction_data.get('aquifer_info'):
                print("✅ Aquifer integration successful!")
                aquifer_info = prediction_data['aquifer_info']
                if aquifer_info.get('success') and aquifer_info.get('nearest_aquifer'):
                    print(f"   - Nearest Aquifer: {aquifer_info['nearest_aquifer']['name']}")
                    print(f"   - Distance: {aquifer_info.get('distance_km')} km")
                    print(f"   - Recommendations: {len(aquifer_info.get('recommendations', []))} items")
                else:
                    print(f"   - Aquifer error: {aquifer_info.get('error', 'Unknown error')}")
            else:
                print("❌ Aquifer integration missing from prediction!")
        else:
            print(f"❌ Prediction endpoint failed: {response.status_code}")
            print(f"   Response: {response.text}")
            
    except Exception as e:
        print(f"❌ Prediction test failed: {str(e)}")
    
    print("\n" + "="*60)
    print("🎯 AQUIFER INTEGRATION TEST COMPLETE")
    print("Next Steps:")
    print("1. Submit an assessment through the frontend with coordinates")
    print("2. Check the Results page for aquifer information display")
    print("3. Verify recommendations are showing correctly")
    print("="*60)

if __name__ == "__main__":
    test_aquifer_integration()

File: test_aquifer_integration.js
/**
 * Test script to verify the aquifer integration works end-to-end
 * This tests the full flow from frontend input to ML service aquifer analysis
 */

const fetch = require('node-fetch');

// Test configuration
const ML_SERVICE_URL = process.env.ML_SERVICE_URL || 'http://localhost:8000';
const BACKEND_URL = process.env.BACKEND_URL || 'http://localhost:5000';

// Sample test data with Bangalore coordinates
const testAssessmentData = {
  roof_area: 100,
  roof_type: "concrete",
  soil_type: "loamy",
  annual_rainfall: 970,
  state: "Karnataka",
  district: "Bangalore Urban",
  latitude: 12.9716,  // Bangalore coordinates
  longitude: 77.5946
};

async function testMLServiceDirect() {
  console.log('\n🧪 Testing ML Service Direct Aquifer Endpoint...');
  
  try {
    const response = await fetch(`${ML_SERVICE_URL}/aquifer-info/${testAssessmentData.latitude}/${testAssessmentData.longitude}`);
    
    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }
    
    const data = await response.json();
    console.log('✅ ML Service Aquifer Response:');
    console.log(`   - Success: ${data.success}`);
    if (data.success && data.nearest_aquifer) {
      console.log(`   - Nearest Aquifer: ${data.nearest_aquifer.name}`);
      console.log(`   - Distance: ${data.distance_km} km`);
      console.log(`   - Type: ${data.aquifer_type}`);
      console.log(`   - Recharge Potential: ${data.recharge_potential}`);
      console.log(`   - Feasibility Score: ${data.feasibility_score}/100`);
      console.log(`   - Assessment: ${data.overall_assessment}`);
      console.log(`   - Recommendations: ${data.recommendations?.length || 0} items`);
    }
    
    return data;
  } catch (error) {
    console.error('❌ ML Service Direct Test Failed:', error.message);
    return null;
  }
}

async function testMLServicePrediction() {
  console.log('\n🧪 Testing ML Service Prediction with Aquifer Data...');
  
  try {
    const response = await fetch(`${ML_SERVICE_URL}/predict`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(testAssessmentData)
    });
    
    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }
    
    const data = await response.json();
    console.log('✅ ML Service Prediction Response:');
    console.log(`   - Potential Harvest: ${data.potential_harvest} L/year`);
    console.log(`   - Tank Volume: ${data.tank_volume} L`);
    console.log(`   - Feasibility: ${data.feasibility}`);
    console.log(`   - Groundwater Level: ${data.groundwater_level}m`);
    
    if (data.aquifer_info) {
      console.log(`   - Aquifer Integration: ✅ Success`);
      console.log(`   - Nearest Aquifer: ${data.aquifer_info.nearest_aquifer?.name}`);
      console.log(`   - Distance to Aquifer: ${data.aquifer_info.distance_km} km`);
    } else {
      console.log(`   - Aquifer Integration: ❌ Missing`);
    }
    
    return data;
  } catch (error) {
    console.error('❌ ML Service Prediction Test Failed:', error.message);
    return null;
  }
}

async function testHealthChecks() {
  console.log('\n🏥 Testing Health Checks...');
  
  // Test ML Service health
  try {
    const mlHealth = await fetch(`${ML_SERVICE_URL}/health`);
    console.log(`✅ ML Service Health: ${mlHealth.status} ${mlHealth.statusText}`);
  } catch (error) {
    console.log(`❌ ML Service Health: ${error.message}`);
  }
  
  // Test Backend health (if available)
  try {
    const backendHealth = await fetch(`${BACKEND_URL}/api/health`);
    console.log(`✅ Backend Health: ${backendHealth.status} ${backendHealth.statusText}`);
  } catch (error) {
    console.log(`ℹ️  Backend Health: ${error.message} (may not have health endpoint)`);
  }
}

async function runTests() {
  console.log('🚀 Starting Aquifer Integration Tests...');
  console.log(`ML Service URL: ${ML_SERVICE_URL}`);
  console.log(`Backend URL: ${BACKEND_URL}`);
  console.log(`Test Location: Bangalore (${testAssessmentData.latitude}, ${testAssessmentData.longitude})`);
  
  // Run all tests
  await testHealthChecks();
  const aquiferData = await testMLServiceDirect();
  const predictionData = await testMLServicePrediction();
  
  console.log('\n📊 Test Summary:');
  console.log(`✅ ML Service Direct Aquifer: ${aquiferData ? 'PASS' : 'FAIL'}`);
  console.log(`✅ ML Service Prediction: ${predictionData ? 'PASS' : 'FAIL'}`);
  console.log(`✅ Aquifer Integration: ${predictionData?.aquifer_info ? 'PASS' : 'FAIL'}`);
  
  if (aquiferData && predictionData?.aquifer_info) {
    console.log('\n🎉 All tests passed! Aquifer integration is working correctly.');
    console.log('\n📋 Next steps:');
    console.log('   1. Test the frontend by submitting an assessment with coordinates');
    console.log('   2. Verify the Results page displays aquifer information');
    console.log('   3. Check that recommendations are showing correctly');
  } else {
    console.log('\n⚠️  Some tests failed. Please check the ML service logs.');
  }
}

// Run the tests
runTests().catch(console.error);